scipy/fft/czt.py

```diff
@@ -0,0 +1,453 @@
+        trnsp[[axis, -1]] = [-1, axis]
+        x = x.transpose(*trnsp)
+        y = ifft(self._Fwk2 * fft(x*self._Awk2, self._nfft))
+        y = y[..., self._yidx] * self._wk2
+        return y.transpose(*trnsp)
+
+    def points(self):
+        """
+        The points at which the Z-transform is computed when calling this
+        `CZT`.
+        """
+        return czt_points(self.m, self.w, self.a)
+
+
+class ZoomFFT(CZT):
+    """
+    Create a zoom FFT transform function.
+
+    This is a specialization of the chirp Z-transform (`CZT`) for a set of
+    equally-spaced frequencies around the unit circle.
+
+    Parameters
+    ----------
+    n : int
+        The size of the signal.
+    fn : array_like, optional
+        A length-2 sequence [f1, f2] giving the frequency range, or a scalar,
+        for which the range [0, fn] is assumed.
+    m : int, optional
+        The number of output points desired.  The default is `n`.
+    Fs : float, optional
+        The sampling frequency. (default=2)
+
+    Returns
+    -------
+    f : ZoomFFT
+        A callable object ``f(x, axis=-1)`` for computing the zoom FFT on `x`.
+
+    Notes
+    -----
+    Sampling frequency is 1/dt, the time step between samples in the
+    signal `x`.  The unit circle corresponds to frequencies from 0 up
+    to the sampling frequency.  The default sampling frequency of 2
+    means that f1, f2 values up to the Nyquist frequency are in the
+    range [0, 1). For f1, f2 values expressed in radians, a sampling
+    frequency of 1/pi should be used.
+
+    Remember that a zoom FFT can only interpolate the points of the existing
+    FFT.  It cannot help to resolve two separate nearby frequencies.
+    Frequency resolution can only be increased by increasing acquisition
+    time.
+
+    Examples
+    --------
+    To plot the transform results use something like the following:
+
+    >>> t = np.linspace(0, 1, 1021)
+    >>> x = np.cos(2*np.pi*15*t) + np.sin(2*np.pi*17*t)
+    >>> f1, f2 = 5, 27
+    >>> transform = ZoomFFT(len(x), [f1, f2], len(x), 1021)
+    >>> X = transform(x)
+    >>> f = np.linspace(f1, f2, len(x))
+    >>> import matplotlib.pyplot as plt
+    >>> plt.plot(f, 20*np.log10(np.abs(X)))
+    >>> plt.show()
+    """
+
+    def __init__(self, n, fn, m=None, Fs=2):
+        m = _validate_sizes(n, m)
+
+        k = arange(max(m, n), dtype=np.min_scalar_type(-max(m, n)**2))
+
+        if np.size(fn) == 2:
+            f1, f2 = fn
+        elif np.size(fn) == 1:
+            f1, f2 = 0.0, fn
+        else:
+            raise ValueError('fn must be a scalar or 2-length sequence')
+
+        self.f1, self.f2, self.Fs = f1, f2, Fs
+
+        scale = ((f2 - f1) * m) / (Fs * (m - 1))
+        a = cmath.exp(2j * pi * f1/Fs)
+        w = cmath.exp(-2j*pi/m * scale)
+        wk2 = np.exp(-(1j * pi * scale * k**2) / m)
+
+        self.w, self.a = w, a
+        self.m, self.n = m, n
+
+        nfft = next_fast_len(n + m - 1)
+        self._Awk2 = (a**-k * wk2)[:n]
```

Comment:
Calculating `a**-k` is likely to cause numerical errors such that the modulus of the result deviates from one. Since we have an explicit expression for `a`, we could calculate the angle first and then project to the complex plane, i.e. something like
```
ak = cmath.exp(-2j * pi * f1/Fs * k)
self._Awk2 = (ak * wk2)[:n]
```
This should yield more accurate results regardless of how python computes powers of c-numbers internally.

Reply:
**For an explanation:**
This means instead of calculating `A**-k`, which is `cmath.exp(2j * pi * f1/Fs)**-k`,
we can pull the outer power as factor into the power of e:
`cmath.exp(-2j * pi * f1/Fs * k)`

This is also expected to be computational more efficient and an optimization described here for the MATLAB implementation:
[SANDIA REPORT (2005) - Chirp Z-Transform Spectral Zoom Optimization with MATLAB](https://core.ac.uk/download/pdf/195151009.pdf)
*See variable definition "aa" in code example of section "Detailed Analysis" (p15) and compare with original code (p11)